compiler/noirc_evaluator/src/ssa/opt/remove_unreachable_instructions.rs

```diff
@@ -0,0 +1,430 @@
+            constrain u1 0 == u1 1, "Index out of bounds"
+            v4 = array_get v0, index u32 0 -> &mut u1
+            v5 = load v4 -> u1
+            jmp b1(v5)
+          b1(v6: u1):
+            v7 = add v6, u1 1
+            jmp b2(v7)
+          b2(v8: u1):
+            v9 = add v8, u1 1
+            return v9
+        }
+        "#;
+        let ssa = Ssa::from_str(src).unwrap();
+        let ssa = ssa.remove_unreachable_instructions();
+
+        assert_ssa_snapshot!(ssa, @r#"
+        acir(inline) predicate_pure fn main f0 {
+          b0():
+            v2 = make_array [] : [&mut u1; 0]
+            constrain u1 0 == u1 1, "Index out of bounds"
+            jmp b1(u1 0)
+          b1(v0: u1):
+            jmp b2(u1 0)
+          b2(v1: u1):
+            return u1 0
+        }
+        "#);
+    }
+
+    #[test]
+    fn removes_unreachable_instructions_from_successors_goes_in_pre_order() {
+        let src = r#"
+        acir(inline) predicate_pure fn main f0 {
+          b0():
+            v0 = make_array [] : [&mut u1; 0]
+            constrain u1 0 == u1 1, "Index out of bounds"
+            v4 = array_get v0, index u32 0 -> &mut u1
+            v5 = load v4 -> u1
+            jmp b2(v5)
+          b1(v8: u1):
+            v9 = add v8, u1 1
+            return v9
+          b2(v6: u1):
+            v7 = add v6, u1 1
+            jmp b1(v7)
+        }
+        "#;
+        let ssa = Ssa::from_str(src).unwrap();
+        let ssa = ssa.remove_unreachable_instructions();
+
+        assert_ssa_snapshot!(ssa, @r#"
+        acir(inline) predicate_pure fn main f0 {
+          b0():
+            v2 = make_array [] : [&mut u1; 0]
+            constrain u1 0 == u1 1, "Index out of bounds"
+            jmp b2(u1 0)
+          b1(v0: u1):
+            return u1 0
+          b2(v1: u1):
+            jmp b1(u1 0)
+        }
+        "#);
+    }
+
+    #[test]
+    fn removes_unreachable_instructions_if_successor_has_other_predecessors() {
+        let src = r#"
+        acir(inline) predicate_pure fn main f0 {
+          b0():
+            constrain u1 0 == u1 1, "Index out of bounds"
+            jmp b1()
+          b1():
+            v1 = add Field 1, Field 2
+            jmpif u1 0 then: b2, else: b3
+          b2():
+            v2 = add Field 1, Field 2
+            jmp b1()
+          b3():
+            v3 = add Field 1, Field 2
+            return v3
+        }
+        "#;
+        let ssa = Ssa::from_str(src).unwrap();
+        let ssa = ssa.remove_unreachable_instructions();
+
+        assert_ssa_snapshot!(ssa, @r#"
+        acir(inline) predicate_pure fn main f0 {
+          b0():
+            constrain u1 0 == u1 1, "Index out of bounds"
+            jmp b1()
+          b1():
+            jmpif u1 0 then: b2, else: b3
+          b2():
+            jmp b1()
+          b3():
+            return Field 0
+        }
+        "#);
+    }
+
+    #[test]
+    fn removes_unreachable_instructions_following_block_with_no_instructions() {
```

Comment:
Why was a block without instructions considered a special case to be tested? Isn't `b1` in the previous test also without instructions?

Reply:
I ended up renaming some tests and adding some comments to them. I left all of them just in case, it doesn't hurt to have a few more tests on this 😊